pkg/ck8s/workload_cluster.go

```diff
@@ -204,6 +205,60 @@ func (w *Workload) GetK8sdProxyForControlPlane(ctx context.Context, options k8sd
 	return nil, fmt.Errorf("failed to get k8sd proxy for control plane")
 }
 
+// GetK8sdProxyForMachine returns a k8sd proxy client for the machine.
+func (w *Workload) GetK8sdProxyForMachine(ctx context.Context, machine *clusterv1.Machine) (*K8sdClient, error) {
+	node := &corev1.Node{}
+	if err := w.Client.Get(ctx, ctrlclient.ObjectKey{Name: machine.Status.NodeRef.Name}, node); err != nil {
+		return nil, fmt.Errorf("failed to get node: %w", err)
+	}
+
+	return w.K8sdClientGenerator.forNode(ctx, node)
+}
+
+func (w *Workload) RefreshMachine(ctx context.Context, machine *clusterv1.Machine, nodeToken *string, upgradeOption *string) (string, error) {
+	request := apiv1.SnapRefreshRequest{}
+	response := &apiv1.SnapRefreshResponse{}
+	optionKv := strings.Split(*upgradeOption, "=")
+	switch optionKv[0] {
+	case "channel":
+		request.Channel = optionKv[1]
+	case "revision":
+		request.Revision = optionKv[1]
+	case "localPath":
+		request.LocalPath = optionKv[1]
+	default:
+		return "", fmt.Errorf("invalid upgrade option: %s", optionKv[0])
+	}
+
+	header := map[string][]string{
+		"node-token": {*nodeToken},
+	}
+
+	err := w.doK8sdRequestForMachine(ctx, machine, http.MethodPost, "1.0/snap/refresh", header, request, response)
+	if err != nil {
+		return "", fmt.Errorf("failed to refresh machine %s: %w", machine.Name, err)
+	}
+
+	return response.ChangeID, nil
+}
+
+func (w *Workload) GetRefreshStatusForMachine(ctx context.Context, machine *clusterv1.Machine, nodeToken *string, changeID *string) (*apiv1.SnapRefreshStatusResponse, error) {
+	request := apiv1.SnapRefreshStatusRequest{}
+	response := &apiv1.SnapRefreshStatusResponse{}
+	request.ChangeID = *changeID
```

Comment:
nit: IMO it's better to keep the `changeID` every where, since we expect a `change ID` from snap (also consider changing the annotations). but I get that we're doing a `refresh` which has a `changeID` and is confusing by itself.